File: skylebank-frontend/src/services/walletService.ts
import { apiClient } from './apiClient';
import type { WalletInfo, WalletBalanceInfo } from '../types/wallet';

export const walletService = {
  /**
   * Retrieves full wallet details for the authenticated user.
   */
  async getWalletDetails(): Promise<WalletInfo> {
    const response = await apiClient.get<WalletInfo>('/api/v1/wallet');
    return response.data;
  },

  /**
   * Retrieves the wallet balance and currency only.
   */
  async getWalletBalance(): Promise<WalletBalanceInfo> {
    const response = await apiClient.get<WalletBalanceInfo>('/api/v1/wallet/balance');
    return response.data;
  },

  /**
   * Performs real-time recipient lookup using account number.
   */
  async lookupRecipient(accountNumber: string): Promise<{ walletNumber: string; fullName: string }> {
    const response = await apiClient.get<{ walletNumber: string; fullName: string }>('/api/v1/wallet/lookup', {
      params: { accountNumber }
    });
    return response.data;
  }
};

File: skylebank-backend/src/main/java/com/skylebank/api/services/WalletService.java
package com.skylebank.api.services;

import com.skylebank.api.dto.WalletResponse;
import com.skylebank.api.models.Wallet;
import com.skylebank.api.repositories.WalletRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Map;

/**
 * Service class managing business logic for User Wallets.
 */
@Service
@RequiredArgsConstructor
@Slf4j
public class WalletService {

    private final WalletRepository walletRepository;

    /**
     * Retrieves full wallet details for a given user email.
     */
    @Transactional(readOnly = true)
    public WalletResponse getWalletByUserEmail(String email) {
        log.info("Fetching wallet details for user email: {}", email);
        Wallet wallet = walletRepository.findByUserEmail(email)
                .orElseThrow(() -> new IllegalArgumentException("Wallet not found for user: " + email));

        return WalletResponse.builder()
                .walletNumber(wallet.getWalletNumber())
                .balance(wallet.getBalance())
                .currency(wallet.getCurrency())
                .walletStatus(wallet.getWalletStatus().name())
                .build();
    }

    /**
     * Retrieves only the balance and currency details for a given user email.
     */
    @Transactional(readOnly = true)
    public Map<String, Object> getWalletBalance(String email) {
        log.info("Fetching balance details for user email: {}", email);
        Wallet wallet = walletRepository.findByUserEmail(email)
                .orElseThrow(() -> new IllegalArgumentException("Wallet not found for user: " + email));

        return Map.of(
                "balance", wallet.getBalance(),
                "currency", wallet.getCurrency()
        );
    }

    /**
     * Looks up recipient full name by wallet number.
     */
    @Transactional(readOnly = true)
    public com.skylebank.api.dto.RecipientLookupResponse lookupRecipient(String walletNumber) {
        log.info("Performing recipient lookup for wallet number: {}", walletNumber);
        Wallet wallet = walletRepository.findByWalletNumber(walletNumber)
                .orElseThrow(() -> new IllegalArgumentException("Account number does not exist"));

        String fullName = wallet.getUser().getFirstName() + " " + wallet.getUser().getLastName();
        return com.skylebank.api.dto.RecipientLookupResponse.builder()
                .walletNumber(wallet.getWalletNumber())
                .fullName(fullName)
                .build();
    }
}
